staging/src/k8s.io/apiserver/pkg/util/proxy/streamtunnel.go

```diff
@@ -0,0 +1,454 @@
+	clone.Header.Del(wsstream.WebSocketProtocolHeader)
+	clone.Header.Del("Sec-Websocket-Key")
+	clone.Header.Del("Sec-Websocket-Version")
+	clone.Header.Del(httpstream.HeaderUpgrade)
+	// Update the http request for an upstream SPDY upgrade.
+	clone.Method = "POST"
+	clone.Body = nil // Remove the request body which is unused.
+	clone.Header.Set(httpstream.HeaderUpgrade, spdy.HeaderSpdy31)
+	clone.Header.Del(httpstream.HeaderProtocolVersion)
+	for i := range spdyProtocols {
+		clone.Header.Add(httpstream.HeaderProtocolVersion, spdyProtocols[i])
+	}
+	return clone
+}
+
+// spdyProtocolsFromWebsocketProtocols returns a list of spdy protocols by filtering
+// to Kubernetes websocket subprotocols prefixed with "SPDY/3.1+", then removing the prefix
+func spdyProtocolsFromWebsocketProtocols(req *http.Request) []string {
+	var spdyProtocols []string
+	for _, protocol := range gwebsocket.Subprotocols(req) {
+		if strings.HasPrefix(protocol, constants.WebsocketsSPDYTunnelingPrefix) && strings.HasSuffix(protocol, constants.KubernetesSuffix) {
+			spdyProtocols = append(spdyProtocols, strings.TrimPrefix(protocol, constants.WebsocketsSPDYTunnelingPrefix))
+		}
+	}
+	return spdyProtocols
+}
+
+var _ http.ResponseWriter = &tunnelingResponseWriter{}
+var _ http.Hijacker = &tunnelingResponseWriter{}
+
+// tunnelingResponseWriter implements the http.ResponseWriter and http.Hijacker interfaces.
+// Only non-upgrade responses can be written using WriteHeader() and Write().
+// Once Write or WriteHeader is called, Hijack returns an error.
+// Once Hijack is called, Write, WriteHeader, and Hijack return errors.
+type tunnelingResponseWriter struct {
+	// w is used to delegate Header(), WriteHeader(), and Write() calls
+	w http.ResponseWriter
+	// conn is returned from Hijack()
+	conn net.Conn
+	// mu guards writes
+	mu sync.Mutex
+	// wrote tracks whether WriteHeader or Write has been called
+	written bool
+	// hijacked tracks whether Hijack has been called
+	hijacked bool
+}
+
+// Hijack returns a delegate "net.Conn".
+// An error is returned if Write(), WriteHeader(), or Hijack() was previously called.
+// The returned bufio.ReadWriter is always nil.
+func (w *tunnelingResponseWriter) Hijack() (net.Conn, *bufio.ReadWriter, error) {
+	w.mu.Lock()
+	defer w.mu.Unlock()
+	if w.written {
+		klog.Errorf("Hijack called after write")
+		return nil, nil, errors.New("connection has already been written to")
+	}
+	if w.hijacked {
+		klog.Errorf("Hijack called after hijack")
+		return nil, nil, errors.New("connection has already been hijacked")
+	}
+	w.hijacked = true
+	klog.V(6).Infof("Hijack returning websocket tunneling net.Conn")
+	return w.conn, nil, nil
+}
+
+// Header is delegated to the stored "http.ResponseWriter".
+func (w *tunnelingResponseWriter) Header() http.Header {
+	return w.w.Header()
+}
+
+// Write is delegated to the stored "http.ResponseWriter".
+func (w *tunnelingResponseWriter) Write(p []byte) (int, error) {
```

Comment:
should we account for this behavior?
> 	// WriteHeader sends an HTTP response header with the provided
	// status code.
	//
	// If WriteHeader is not called explicitly, the first call to Write
	// will trigger an implicit WriteHeader(http.StatusOK).
	// Thus explicit calls to WriteHeader are mainly used to
	// send error codes or 1xx informational responses.

https://go.dev/src/net/http/server.go?h=WriteHeader

Reply:
oh, nvm, this is delegated, so the delegated ResponseWriter will do it for us